### ISSUE TITLE
Improve environment detection for Python

### ISSUE THREAD

**Issue by symwell** [getappmap/appmap-python#194]
Detection of the environment (prod vs dev) in Python can be improved. It currently detects reliably if Flask or Django are in dev but not if they're in prod.  If added in prod without setting the APPMAP env var at all, it says it records by default and when I issue requests they're not recorded.

We believe it should be changed to detect if it's in prod (and to not record by default in prod) and to then end up in a state where we can say in the docs: adding AppMap doesn't affect production except if added manually for Java.

Basically
```
if (dev or test or some_APPMAP_flag=true) then enable
```
It's hard to detect `if production` and it's dangerous.

This blocks https://github.com/getappmap/applandinc.github.io/issues/906

**Comment by symwell** [comment]
Clarification:

In Flask, it says it's recording by default but doesn't record by default.
```
PYTHONPATH=/home/test/src/appmap-python FLASK_APP=app.py flask run -p 8000
[2022-11-28 10:10:04,098] INFO appmap._implementation.detect_enabled: AppMap recording is enabled because will record by default.
```

In Django (with settings.py changed to have `DEBUG = False`) it says it's recording by default, then says it won't record by default, and doesn't record by default.
```
PYTHONPATH=/home/test/src/appmap-python:$PWD/init python manage.py runserver 127.0.0.1:8000
[2022-11-28 10:21:10,768] INFO appmap._implementation.detect_enabled: AppMap recording is enabled because will record by default.
[2022-11-28 09:21:11,180] INFO appmap._implementation.detect_enabled: AppMap recording is not enabled because couldn't load DJANGO_SETTINGS_MODULE={os.environ['DJANGO_SETTINGS_MODULE']}.
```

**Comment by symwell** [comment]
On [branch](https://github.com/getappmap/appmap-python/compare/sw/improve_environment_detection_for_python__detect_enabled_in_env) `sw/improve_environment_detection_for_python__detect_enabled_in_env`, if all test files are removed except `test_flask.py` and `test_test_frameworks.py`, the `test_test_frameworks.py` tests fail:
```
appmap-python/appmap/test$ ls -1
appmap_test_base.py
conftest.py
data
helpers.py
__init__.py
normalize.py
__pycache__
test_flask.py
test_test_frameworks.py
web_framework.py
```
If the tests of `test_flask.py` are all commented out except `test_framework_metadata` the tests fail.

`test_framework_metadata` running causes the failure in `test_test_frameworks.py`

**Comment by symwell** [comment]
For recording-by-default to work properly, `_enabled` needs to be `False` in prod for `"pytest"` and `"unittest"`.
Else it says
```
src/appmap-python/appmap/test/data/django$ PYTHONPATH=/home/test/src/appmap-python:$PWD/init python manage.py runserver 127.0.0.1:8000
[2022-12-01 09:25:16,539] INFO appmap._implementation.env: AppMap recording is enabled because pytest records by default.
```
With the logic on `master` it was set to `True` for `"pytest"` and `"unittest"` even when running in prod and remote recording and request recording were disabled in prod.  This worked as expected for test but not for prod. It was set through [this line](https://github.com/getappmap/appmap-python/blob/master/appmap/_implementation/detect_enabled.py#L141) in `_detect_should_enable` after a call to `any_enabled`
```
         if recording_method in RECORDING_METHODS:
            return True, logging.INFO, self._log_prefix(True, f"{recording_method} records by default")
```
If a new function is added called `enabled_for` with the following logic in `env.py`, to call it from test or prod accordingly -- to call for `"pytest"` in test and for `"record"` or `"remote"` in prod...
```
--- a/appmap/_implementation/env.py
+++ b/appmap/_implementation/env.py
@@ -52,6 +52,11 @@ class Env(metaclass=_EnvMeta):
         output_dir = Path(self.get("APPMAP_OUTPUT_DIR", "tmp/appmap"))
         self._output_dir = output_dir.resolve()
 
+    def enable_for(self, recording_method):
+        self._detected_for_method = {}
+        #self._enabled = True
+        self._enabled = self.should_enable(recording_method)
```
... and then it's used in a testcase that fails
```
--- a/appmap/test/test_test_frameworks.py
+++ b/appmap/test/test_test_frameworks.py
@@ -10,10 +10,13 @@ import pytest
 from appmap.test.helpers import DictIncluding
 
 from .._implementation import web_framework
+from .._implementation.env import Env
 from .normalize import normalize_appmap
 
 
 def test_unittest_runner(testdir):
+    Env.current.set("APPMAP", "true")
+    Env.current.enable_for("pytest")
     testdir.run(sys.executable, "-m", "unittest", "-vv")
```
then the testcase still fails.

I don't understand this part, regardless of how `Env.current._enabled` is set initially.
Whether `APPMAP` is set to true or false, this testcase always fails.  As if something didn't get enabled correctly or the instrumentation didn't trigger.

**Comment by symwell** [comment]
After this [fix](https://github.com/getappmap/appmap-python/commit/06bb56bd216d77ddc451fe85394d552e93d22f76) now seeing something different on [CI](https://app.travis-ci.com/github/getappmap/appmap-python):
- Python 3.10 breaks with `ModuleNotFoundError: No module named 'setuptools.command.build'`
- Python 3.9 and 3.7 break with `E       AttributeError: 'HttpResponse' object has no attribute 'headers'`
- Python 3.8 breaks with failures in `appmap/test/test_recording.py:`

**Comment by symwell** [comment]
The current state of this work is on branch `sw/improve_environment_detection_for_python__detect_enabled_in_env` at https://github.com/getappmap/appmap-python/compare/sw/improve_environment_detection_for_python__detect_enabled_in_env?expand=1

A summary:
- DetectEnabled was embedded in Env because DetectEnabled used `os.environ` while Env used a copy of the env.  This was leading to DetectEnabled not detecting the same state represented in Env.
- `Env.should_enable` was added to detect if AppMap should be enabled by default based on the context it's called from.  It needs to know it's called from `pytest` to be enabled by default for `pytest`, else it would get enabled by default in prod.
